### PR TITLE
Add options config for backup/restore commands

### PR DESCRIPTION
**Pull request by timkelty**
### Description
The PR allows `\craft\config\GeneralConfig::backupCommand` and `\craft\config\GeneralConfig::restoreCommand` to be an options array to configure how the commands are generated.

#### Goals
- make small tweaks easy (eg ignore a single table). Currently this can only be done via event.
- make it easy to opt-in to postgres's archive format (possibly make this the default for v5?)
- make it clear when things are going to be replaced or merged (eg `ignore-tables`)
- provide a way to tap into command for modification
- be non-breaking

Valid values include:

```
<?php
use craft\config\GeneralConfig;
use mikehaertl\shellcommand\Command as ShellCommand;

return GeneralConfig::create()
    ->restoreCommand([
        'archiveFormat' => true, // pgsql-only

        // example: add an arg
        'callback' => fn(ShellCommand $command) => $command->addArg('--verbose'),
    ])
    ->backupCommand([
        'ignoreTables' => ['foo', 'bar'],
        'archiveFormat' => true, // pgsql-only

        // example: remove an arg
        'callback' => function (ShellCommand $command) {
            $command->setArgs(str_replace('--dump-date', '', $command->getArgs()));

            return $command;
        },
    ]);
```

#### etc…
- I ended up leveraging `mikehaertl\shellcommand\Command` because it was there and were already using it. Easier than stitching strings together.
- To keep backward compat, `ignoreTables` can either come from the `EVENT_BEFORE_CREATE_BACKUP` or from the new `ignoreTables` option. If both are present, the event will take precedence.
- The `callback` could easily be an event instead, but it seemed like it might be a pain to add an event listener just to add an additional arg.

### CODE REVIEW

**Comment by brandonkelly** [comment]
> The options array is keyed by the db driver name

This feels like unnecessary complexity to me. No project uses both.

**Comment by timkelty** [comment]
@brandonkelly yep, already nixed it.

File: src/db/pgsql/Schema.php [hunk 1]
```diff
@@ -312,14 +300,4 @@ protected function getIndexInformation(TableSchema $table): array
...
-
-    /**
-     * Returns the PGPASSWORD command for backup/restore actions.
-     *
-     * @return string
-     */
-    private function _pgpasswordCommand(): string
-    {
-        return Platform::isWindows() ? 'set PGPASSWORD="{password}" && ' : 'PGPASSWORD="{password}" ';
-    }
```
Comment (timkelty) [inline src/db/pgsql/Schema.php:324]: Moved to `\craft\db\DbShellCommand`

File: src/db/mysql/Schema.php [hunk 2]
```diff
@@ -382,47 +318,4 @@ protected function findConstraints($table): void
...
-            $contents .= PHP_EOL . 'ssl_key=' . $this->db->attributes[PDO::MYSQL_ATTR_SSL_KEY];
-        }
-
-        FileHelper::writeToFile($this->tempMyCnfPath, '');
-        // Avoid a “world-writable config file 'my.cnf' is ignored” warning
-        chmod($this->tempMyCnfPath, 0600);
-        FileHelper::writeToFile($this->tempMyCnfPath, $contents, ['append']);
-
-        return $this->tempMyCnfPath;
-    }
```
Comment (timkelty) [inline src/db/mysql/Schema.php:427]: Moved to `\craft\db\DbShellCommand`

File: src/db/mysql/Schema.php [hunk 3]
```diff
@@ -152,79 +155,13 @@ public function createColumnSchemaBuilder($type, $length = null): ColumnSchemaBu
...
-            ' {database}';
-
-        $dataDump = 'mysqldump' .
-            $defaultArgs .
-            ' --no-create-info' .
-            ' ' . implode(' ', $ignoreTableArgs) .
-            ' {database}' .
-            ' >> "{file}"';
-
-        return $schemaDump . ' && ' . $dataDump;
```
Comment (timkelty) [inline src/db/mysql/Schema.php:227]: Moved to `\craft\db\mysql\BackupCommand`

File: src/db/pgsql/Schema.php [hunk 4]
```diff
@@ -116,29 +125,13 @@ public function getLastInsertID($sequenceName = ''): string
...
-            ' --port={port}' .
-            ' --username={user}' .
-            ' --if-exists' .
-            ' --clean' .
-            ' --no-owner' .
-            ' --no-privileges' .
-            ' --no-acl' .
-            ' --file="{file}"' .
-            ' --schema={schema}' .
-            ' ' . implode(' ', $ignoredTableArgs);
```
Comment (timkelty) [inline src/db/pgsql/Schema.php:141]: Moved to `\craft\db\pgsql\RestoreCommand`

File: src/db/pgsql/Schema.php [hunk 5]
```diff
@@ -148,14 +141,9 @@ public function getDefaultBackupCommand(?array $ignoreTables = null): string
...
     public function getDefaultRestoreCommand(): string
     {
-        return $this->_pgpasswordCommand() .
-            'psql' .
-            ' --dbname={database}' .
-            ' --host={server}' .
-            ' --port={port}' .
-            ' --username={user}' .
-            ' --no-password' .
-            ' < "{file}"';
```
Comment (timkelty) [inline src/db/pgsql/Schema.php:158]: Moved to `\craft\db\pgsql\RestoreCommand`

File: src/db/DbShellCommand.php [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        return $this->getCommand()->getExecCommand();
+    }
+
+    /**
+     * Creates a temporary my.cnf file based on the DB config settings.
+     *
+     * @return string The path to the my.cnf file
+     * @throws ErrorException
+     */
+    protected function createDumpConfigFile(): string
```
Comment (angrybrad) [inline src/db/DbShellCommand.php:44]: Worth renaming to `createMySQLDumpConfigFile()` so it's more explicit?

Comment (timkelty) [reply src/db/DbShellCommand.php:44]: Probably? – only reason I didn't is it seems to have logic in it for handling pgsql, oddly enough.

Comment (angrybrad) [reply src/db/DbShellCommand.php:44]: Hrm... that feels unnecessary, given the first thing it's going to do is throw an exception if it's not MySQL. Probably some leftover code.

Comment (timkelty) [reply src/db/DbShellCommand.php]: Renamed and removed pg conditions

**Comment by brandonkelly** [comment]
@timkelty After reading https://github.com/craftcms/cms/issues/14648#issuecomment-2057541574 I’m wondering if it would make more sense to just add a bunch of new top-level config settings for this. Brad is suggesting using a multi-env config in response, which we are sortof moving away from in favor of environment variables, but it doesn’t seem like it will be possible to use environment variables here as currently implemented. (Besides pulling them in manually from `config/general.php` via `App::env()`.)

**Comment by timkelty** [comment]
Closed in favor of https://github.com/craftcms/cms/pull/14897

**Pull request closed without merging**